File: day07/aoc2025_07.scala
import scala.collection.mutable
object Main07 {
    def main(args: Array[String]): Unit = {
        val lines = scala.io.Source.fromFile("day07/input07").getLines().toList
        var beam_indices = mutable.Map[Int, Long](lines(0).indexOf('S') -> 1)
        var part1 = 0
        for (line <- lines.tail)
            val beam_indices_new = mutable.Map[Int, Long]().withDefaultValue(0)
            for ((idx, count) <- beam_indices)
                line(idx) match
                    case '^' =>
                        beam_indices_new(idx - 1) += count
                        beam_indices_new(idx + 1) += count
                        part1 += 1
                    case '.' =>
                        beam_indices_new(idx) += count
            beam_indices = beam_indices_new
        println("Part 1: " + part1)
        println("Part 2: " + beam_indices.values.sum)
    }
}

File: day07/aoc2025_07.ts
import { readFileSync } from "node:fs";

const lines = readFileSync('day07/input07', 'utf-8').trim().split("\n")
let beam_indices = new Map<number, number>()
beam_indices.set(lines[0].indexOf('S'), 1)

let part1 = 0
for (const line of lines.slice(1)) {
    let beam_indices_new = new Map<number, number>()
    for (const [idx, count] of beam_indices.entries()) {
        switch (line[idx]) {
            case '^':
                beam_indices_new.set(idx - 1, (beam_indices_new.get(idx - 1) ?? 0) + count)
                beam_indices_new.set(idx + 1, (beam_indices_new.get(idx + 1) ?? 0) + count)
                part1 += 1
                break;
            case '.':
                beam_indices_new.set(idx, (beam_indices_new.get(idx) ?? 0) + count)
                break;
        }
    }
    beam_indices = beam_indices_new
}
const part2 = Array.from(beam_indices.values()).reduce((x, y) => x + y)
console.log('Part 1:', part1)
console.log('Part 2:', part2)
